File: backend/app/services/stt_channels.py
"""Speechmatics realtime channel diarization (source labeling).

Docs: https://docs.speechmatics.com/speech-to-text/realtime/realtime-diarization
SaaS max 2 channels. Audio is sent with AddChannelAudio, not mixed PCM.
"""

from __future__ import annotations

from typing import Any, Optional

COPILOT_CHANNEL_MODE = "copilot_channels"
ALLOWED_CHANNEL_LABELS = ("prospect", "rep")


def parse_channel_labels(raw: Optional[str]) -> list[str]:
    if not raw or not str(raw).strip():
        return ["prospect", "rep"]
    labels: list[str] = []
    for part in str(raw).split(","):
        p = part.strip().lower()
        if p in ALLOWED_CHANNEL_LABELS and p not in labels:
            labels.append(p)
        if len(labels) == 2:
            break
    return labels or ["prospect"]


def transcription_diarization_for_mode(
    mode: str,
    *,
    channel_labels: Optional[list[str]] = None,
) -> dict[str, Any]:
    """Diarization keys for Speechmatics transcription_config (not the full config)."""
    if mode == COPILOT_CHANNEL_MODE:
        labels = channel_labels or ["prospect", "rep"]
        return {
            "diarization": "channel",
            "channel_diarization_labels": labels,
        }
    if mode in ("enroll", "copilot"):
        return {"diarization": "speaker"}
    return {"diarization": "none"}


def sm_add_channel_audio(channel: str, data_b64: str) -> dict[str, str]:
    return {
        "message": "AddChannelAudio",
        "channel": channel,
        "data": data_b64,
    }


def client_text_to_sm_item(
    data: dict[str, Any],
    allowed_channels: list[str],
) -> dict[str, Any] | str | None:
    msg_type = data.get("type") or data.get("message")
    if msg_type == "AddChannelAudio":
        channel = data.get("channel")
        raw = data.get("data")
        if channel not in allowed_channels or not isinstance(raw, str) or not raw:
            return None
        return sm_add_channel_audio(str(channel), raw)
    if msg_type in ("Finalize", "ForceEndOfUtterance"):
        item: dict[str, Any] = {"message": "ForceEndOfUtterance"}
        channel = data.get("channel")
        if channel in allowed_channels:
            item["channel"] = channel
        return item
    if msg_type == "CloseStream":
        return "close"
    return None


def speechmatics_audio_channel(data: dict[str, Any]) -> Optional[str]:
    ch = data.get("channel")
    if isinstance(ch, str) and ch in ALLOWED_CHANNEL_LABELS:
        return ch
    return None


def accepts_client_pcm_bytes(mode: str) -> bool:
    """Channel sessions must not mix raw PCM into AddChannelAudio streams."""
    return mode != COPILOT_CHANNEL_MODE

File: chrome-extension/lib/stt-channels.js
/**
 * Source-labeled STT for tab copilot.
 * Speechmatics realtime channel diarization:
 * https://docs.speechmatics.com/speech-to-text/realtime/realtime-diarization
 */

export const COPILOT_CHANNEL_MODE = 'copilot_channels';
export const CHANNEL_PROSPECT = 'prospect';
export const CHANNEL_REP = 'rep';

export function parseChannelLabels(raw) {
  if (raw == null || String(raw).trim() === '') return [CHANNEL_PROSPECT, CHANNEL_REP];
  const out = [];
  for (const part of String(raw).split(',')) {
    const p = part.trim().toLowerCase();
    if ((p === CHANNEL_PROSPECT || p === CHANNEL_REP) && !out.includes(p)) {
      out.push(p);
    }
    if (out.length === 2) break;
  }
  return out.length ? out : [CHANNEL_PROSPECT];
}

export function isCoachableChannel(audioChannel) {
  return audioChannel !== CHANNEL_REP;
}

export function encodeChannelAudio(channel, pcmBuffer) {
  const bytes = new Uint8Array(pcmBuffer);
  let binary = '';
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return JSON.stringify({
    type: 'AddChannelAudio',
    channel,
    data: btoa(binary),
  });
}

/** Sets Speechmatics channel-diarization query params on the live STT URL. */
export function applyChannelLabelsToLiveUrl(wsUrl, labels) {
  const parsed = Array.isArray(labels)
    ? parseChannelLabels(labels.join(','))
    : parseChannelLabels(labels);
  const url = new URL(wsUrl);
  url.searchParams.set('mode', COPILOT_CHANNEL_MODE);
  url.searchParams.set('channel_labels', parsed.join(','));
  return url.toString();
}
